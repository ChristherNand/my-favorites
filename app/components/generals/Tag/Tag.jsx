import React, {useState, useRef} from 'react';
import {TagPropType} from '../../../propsTypes';
import {connect} from 'react-redux';
import "./Tag.css"

import Icon from '../Icon';
import {updateTag} from "../../../state/actions";

const mapDispachToProps = dispatch => ({
        updateTag: (id, name, color) => dispatch(updateTag(id, name, color)),
      })

export function Tag(props){
    const inputEl = useRef(null);
    const [isEdit, setIsEdit] = useState(false);
    const {color = "0396A6", onClose, name = "", updateTag, id, updateDisable = false} = props;

    function activeEdit(){
      if(updateDisable){return};
      setIsEdit(true);
    }

    function hideEdit(){
      setIsEdit(false);
    }
    
    function checkName(event){
      if(event.key == "Enter" && inputEl.current.value !== ""){
          updateTag(id, inputEl.current.value, color);
      }
    }

    return  <div className="Tag --ellipsis" style={{"background": `#${color}`}} onBlur={hideEdit}>
              <Icon name="close" onClick={onClose} className={"close"} pointer={true}/>
              {!isEdit && <span className="pointer" onDoubleClick={activeEdit}>{name}</span>}
              {isEdit && <input className="editBox" placeholder={name} ref={inputEl} autoFocus onKeyPress={checkName}/>}
            </div>

}

Tag.propType = TagPropType;

export default connect(null, mapDispachToProps)(Tag);
