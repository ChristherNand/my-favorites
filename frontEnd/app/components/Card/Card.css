.card{
  --num-cards: 3;
  --icon-size: 18px;
  --cards-margin: 10px;
  --elemts-margin-top: 10px;

  display: flex;
  padding: 10px;
  font-size: 16px;
  color: var(--gray);
  align-items: center;
  box-sizing: border-box;
  flex-direction: column;
  justify-content: center;
  background: var(--white);
  margin: calc(var(--cards-margin) / 2);
  width: calc(100% / var(--num-cards) - var(--cards-margin));
}

.card div:empty{
  display: none;
}

.linkTitle{
  width: 100%;
  text-decoration: none;
}

.description{
  font-size: 14px;
  max-height: 88px;
  white-space: pre-line;
}

.description, .card_options, .contTags{
  margin-top: var(--elemts-margin-top);
}

.contTags{
  width: 100%;
}

.material-icons.iconHover{
  opacity: 0.8;
  color: var(--blue);
  font-size: var(--icon-size);
}

.iconHover:hover{
  opacity: 1;
}

.loading{
  width: 24px;
  margin: 6px;
  height: 24px;
  display: flex;
  position: relative;
  justify-content: center;
  align-items: center;
}

.loading::before{
  content: '';
  width: 28px;
  height: 28px;
  position: absolute;
  border-radius: 50%;
  pointer-events: none;
  border: 4px solid var(--light-gray-4);
}

.loading::after{
  content: "";
  position: absolute;
  top: 48%;
  left: -6px;
  width: 4px;
  height: 4px;
  margin: -2px 0 0 0;
  border-radius: 50%;
  background: var(--gray);
  transform-origin: 18px 50%;
  animation-duration: 5s, 0.4s;
  animation-fill-mode: forwards;
  animation-iteration-count: 1, infinite;
  /* animation-iteration-count: infinite; */
  animation-timing-function: ease, linear;
  /* animation-name: anim-effect-simo-3; */
  animation-name: anim-effect-simo-2, anim-effect-simo-3;
}

@keyframes anim-effect-simo-2 {
	0%, 5%, 95%, 100% {
		opacity: 0;
	}
	10%, 90% {
		opacity: 1;
	}
}

@keyframes anim-effect-simo-3 {
	100% {
		-webkit-transform: rotate3d(0, 0, 1, 360deg);
		transform: rotate3d(0, 0, 1, 360deg);
	}
}

