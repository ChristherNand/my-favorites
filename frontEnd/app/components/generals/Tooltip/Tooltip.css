.Tooltip{
  left: 50%;
  width: 200px;
  font-size: 14px;
  color: var(--white);
  padding: 6px 10px;
  position: absolute;
  border-radius: 6px;
  margin-left: -100px;
  word-break: break-all;
  background: var(--blue);
  bottom: calc(100% + 8px);
  box-sizing: border-box;
}

.Tooltip::before{
  content: "";
  width: 0;
  height: 0;
  left: 10px;
  bottom: -5px;
  position: absolute;
  border-top: 6px solid var(--blue);
  border-left: 6px solid transparent;
  border-right: 6px solid transparent;
}
