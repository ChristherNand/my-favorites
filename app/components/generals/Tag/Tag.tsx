import React, { useState, useRef, Dispatch } from 'react';
import './Tag.css';

import Icon from '../Icon';
import { updateTagAsyncAction } from '../../../state/actions';
import { iTag } from '../../../interfaces';
import { useAppDispatch } from '../../../hooks/redux';

interface iProps {
  id?: string | number;
  name: string;
  color?: string;
  isUpdateDisable?: boolean;
  onClose: (tag: iTag) => void;
}

const Tag: React.FC<iProps> = (props) => {
  const inputEl = useRef(null);
  const [isEdit, setIsEdit] = useState(false);
  const dispatchUpdateTag: Dispatch<any> = useAppDispatch();

  const updateTag = (id: string | number, name: string) =>
    dispatchUpdateTag(updateTagAsyncAction(id, name));

  const { color = '0396A6', onClose, name = '', id, isUpdateDisable = false } = props;

  const activeEdit = () => {
    if (isUpdateDisable) {
      return;
    }
    setIsEdit(true);
  };

  const hideEdit = () => setIsEdit(false);

  const checkName = (event) => {
    if (event.key == 'Enter' && inputEl.current.value !== '') {
      updateTag(id, inputEl.current.value);
    }
  };

  return (
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    <div
      className="Tag --ellipsis"
      style={{ background: `#${color}` }}
      data-test="container"
      onBlur={hideEdit}
    >
      <Icon
        name="close"
        pointer={true}
        data-test="Icon"
        className={'close'}
        onClick={() => onClose({ id, name, color })}
      />
      {!isEdit && (
        <button
          className={isUpdateDisable ? 'noCursor' : ''}
          onDoubleClick={activeEdit}
          data-test="tag-name"
        >
          {name}
        </button>
      )}
      {isEdit && (
        <input
          // eslint-disable-next-line jsx-a11y/no-autofocus
          autoFocus
          className="editBox"
          placeholder={name}
          ref={inputEl}
          onKeyPress={checkName}
          data-test="input-edit"
        />
      )}
    </div>
  );
};

export default Tag;
