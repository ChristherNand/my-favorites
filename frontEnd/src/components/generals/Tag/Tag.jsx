import React, {useState, useRef} from 'react';
import "./Tag.css"

import Icon from '../Icon';
import {updateTag} from "../../../utils/ServerRequest";
import {useStateValueCtx} from "../../../context/Tags.contex";

function Tag(props){
    const inputEl = useRef(null);
    const [isEdit, setIsEdit] = useState(false);
    const {color = "0396A6", onClose, name = ""} = props;
    /* const [dataCtx, dispatchCtx] = useStateValueCtx(); */

    function activeEdit(){
        setIsEdit(true);
    }

    function hideEdit(){
        setIsEdit(false);
    }
    
    function checkName(event){
        if(event.key == "Enter" && inputEl.current.value !== ""){
            putTag(inputEl.current.value)
        }
    }

    async function putTag(name){
        try{
            const {status, data} = await updateTag(props.id, name);
            if(status == "updated"){
                /* dispatchCtx({ type: "updateTags", id: data.id, newName: data.name}) */
            }
        }catch(error){
            console.error("error", error)
        }
    }

    return  <div className="Tag --ellipsis" style={{"background": `#${color}`}} onBlur={hideEdit}>
                <Icon name="close" onClick={onClose} className={"close"} pointer={true}/>
                {!isEdit && <span className="pointer" onDoubleClick={activeEdit}>{name}</span>}
                {isEdit && <input className="editBox" placeholder={name} ref={inputEl} autoFocus onKeyPress={checkName}/>}
            </div>

}

export default Tag;