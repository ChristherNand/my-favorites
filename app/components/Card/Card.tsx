import React, { Dispatch, useRef } from 'react';
import './Card.css';

import Icon from '../generals/Icon';
import Tag from '../generals/Tag';
import Tooltip from '../generals/Tooltip';
import { removeTagLinkAsyncAction, removeLinkAsyncAction } from '../../state/actions';
import { iLink } from '../../interfaces';
import { useAppDispatch } from '../../hooks/redux';

const fakeText = `Donec in venenatis metus. Suspendisse potenti. Cras ultricies turpis sit amet massa suscipit, in egestas sapien finibus.`;

const Card: React.FC<iLink> = ({ id = '0', title, url, tags = [], description = fakeText }) => {
  const containerUrl = useRef<HTMLButtonElement>(null);
  const [isContHover, setIsContHover] = React.useState(false);

  const dispatchRemoveLink: Dispatch<any> = useAppDispatch();
  const dispatchRemoveTag: Dispatch<any> = useAppDispatch();
  const removeLink = () => dispatchRemoveLink(removeLinkAsyncAction(id as string));
  const removeTag = (tagId: string) => dispatchRemoveTag(removeTagLinkAsyncAction(id as string, tagId));

  const editLink = () => {
    // Todo edit link method backed
    console.log('editLink...', id);
  }

  return (
    <div className="card">
      <a className="linkTitle" href={url} target="_blank" data-test="link-url" rel="noreferrer">
        <h2 className="title" data-test="title">
          {title}
        </h2>
      </a>

      <div className="contDescription">
        <button
          className="description"
          ref={containerUrl}
          onMouseEnter={() => setIsContHover(true)}
          onMouseLeave={() => setIsContHover(false)}
          data-test="description"
        >
          {description}
        </button>
        <Tooltip
          text={description}
          calcHeight={true}
          hover={isContHover}
          parentRef={containerUrl}
          data-test="cp-tooltip"
        />
      </div>

      <div className="card__cont contTags --flex --wrap">
        {tags.map((tag, index) => (
          <Tag
            data-test="cp-tag"
            key={`${id}-${index}`}
            isUpdateDisable={true}
            onClose={() => removeTag(tag.id as string)}
            {...tag}
          />
        ))}
      </div>
      <div className="card_options --flex">
        <div className="loading">
          <Icon
            name="edit"
            onClick={editLink}
            className={'edit iconHover'}
            pointer={true}
            data-test="btn-edit"
          />
        </div>
        <div className="loading">
          <Icon
            name="close"
            onClick={removeLink}
            className={'close iconHover'}
            pointer={true}
            data-test="btn-remove"
          />
        </div>
      </div>
    </div>
  );
}

export default Card;
