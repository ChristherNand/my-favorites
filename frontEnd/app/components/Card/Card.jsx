import React, {useRef, useState} from 'react';
import {connect} from 'react-redux';
import {CardPropType} from '../../propsTypes';
import './Card.css';

import Icon from '../generals/Icon';
import Tag from '../generals/Tag';
import Tooltip from '../generals/Tooltip';
import {removeTagLink} from "../../state/actions";

const mapDispachToProps = dispatch => ({
    removeTagLink: (linkId, tagId) => dispatch(removeTagLink(linkId, tagId)),
  })

const fakeText = `Donec in venenatis metus. Suspendisse potenti. Cras ultricies turpis sit amet massa suscipit, in egestas sapien finibus. 
  Quisque id massa pretium felis auctor mattis. Donec elementum est ut nisi finibus auctor.`;

function Card(props) {
  const {id = 0, title, url, tags = [], description = fakeText} = props;
  const containerUrl = useRef(null);
  const [isContHover, setIsContHover] = useState(false);

  function removeTag(tagId) {
    props.removeTagLink(props.id, tagId);
  }
  
  function removeCard() {
    console.log("cardId", props.id)
  }

  return  <div className="card">
            <a className="linkTitle" href={url} target="_blank">
              <h2 className="title">{title}</h2>
            </a>

            <div className="contDescription">
              <p className="description" ref={containerUrl} onMouseEnter={() => setIsContHover(true)} onMouseLeave={() => setIsContHover(false)}>
                {description}
              </p>
              <Tooltip text={description} calcHeight={true} hover={isContHover} parentRef={containerUrl}/>
            </div>

            <div className="card__cont contTags --flex --wrap">{
              tags.map((tag, index) => {
                return <Tag key={`${id}-${index}`}
                            updateDisable={true} {...tag}
                            onClose={() => removeTag(tag.id)}
                        />
              })}
            </div>
            <div className="card_options --flex">
              <div className="loading">
                <Icon name="edit" onClick={removeCard} className={"edit iconHover"} pointer={true}/>
              </div>
              <div className="loading">
                <Icon name="close" onClick={removeCard} className={"close iconHover"} pointer={true}/>
              </div>
            </div>
          </div>
}

Card.propTypes = CardPropType

export default connect(null, mapDispachToProps)(Card);
