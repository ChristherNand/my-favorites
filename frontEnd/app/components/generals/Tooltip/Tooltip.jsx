import React, {useEffect, useState} from 'react';
import PropTypes from "prop-types";

import "./Tooltip.css";

function Tooltip (props){
  const {text, parentRef, hover} = props;
  const [show, setShow] = useState(false);

  useEffect(() => {
    const {clientWidth, scrollWidth} = parentRef.current;
    if(hover && clientWidth < scrollWidth){
      setShow(true);
    }else{
      setShow(false);
    }
  }, [hover])

  if(!show) {return null}

  return  <div className="Tooltip">
              {text}
            </div>
}

Tooltip.propType = {
  test: PropTypes.string.isRequired,
  parentRef: PropTypes.any.isRequired,
};

export default Tooltip;
