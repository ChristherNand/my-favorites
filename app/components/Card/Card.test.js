import React from 'react';
import { shallow } from 'enzyme';

import { Card } from './Card';
import Tag from '../generals/Tag';
import Tooltip from '../generals/Tooltip';

var Component;
const spyObj = {
  removeLink: jest.fn(),
  removeTagLink: jest.fn(),
}
const data = {id: 1991, 
              url: 'url test', 
              title: 'title test', 
              tags: [{id: 123}, {id: 456}, {id: 789}],
              description: 'fake description'}

beforeEach(() => {});

test('take snapshot', () => {
  Component = shallow(<Card {...data}/>);
  expect(Component).toMatchSnapshot();
});

test('render basic info', () => {
  const Component = shallow(<Card {...data}/>);
  expect(Component.find('h2').text()).toBe(data.title);
  expect(Component.find('.linkTitle').props().href).toBe(data.url);
  expect(Component.find('.description').text()).toBe(data.description);
});

test('render tags', () => {
  const Component = shallow(<Card {...data}/>);
  expect(Component.find(Tag)).toHaveLength(data.tags.length);
});

test('tooltip should be hidden', () => {
  const Component = shallow(<Card {...data}/>);
  expect(Component.find(Tooltip).props().hover).toBe(false);
});

test('show tooltip tooltip', () => {
  const Component = shallow(<Card {...data}/>);
  Component.find('.description').simulate('mouseEnter');
  expect(Component.find(Tooltip).props().hover).toBe(true);
});

test('trigger remove link', () => {
  const Component = shallow(<Card {...data} {...spyObj}/>);
  Component.find('.close').simulate('click');
  expect(spyObj.removeLink).toHaveBeenLastCalledWith(data.id);
});

test('trigger edit link', () => {
  const Component = shallow(<Card {...data} {...spyObj}/>);
  Component.find('.close').simulate('click');
  expect(spyObj.removeLink).toHaveBeenLastCalledWith(data.id);
});

test('remove a tag', () => {
  const Component = shallow(<Card {...data} {...spyObj}/>);
  Component.find(Tag).at(0).prop('onClose')();
  expect(spyObj.removeTagLink).toHaveBeenLastCalledWith(data.id, data.tags[0].id);
});

