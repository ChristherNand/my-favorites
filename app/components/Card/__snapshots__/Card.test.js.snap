// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`take snapshot 1`] = `
<div
  className="card"
>
  <a
    className="linkTitle"
    href="url test"
    target="_blank"
  >
    <h2
      className="title"
    >
      title test
    </h2>
  </a>
  <div
    className="contDescription"
  >
    <p
      className="description"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      fake description
    </p>
    <Tooltip
      calcHeight={true}
      hover={false}
      parentRef={
        Object {
          "current": null,
        }
      }
      text="fake description"
    />
  </div>
  <div
    className="card__cont contTags --flex --wrap"
  >
    <Connect(Tag)
      id={123}
      key="1991-0"
      onClose={[Function]}
      updateDisable={true}
    />
    <Connect(Tag)
      id={456}
      key="1991-1"
      onClose={[Function]}
      updateDisable={true}
    />
    <Connect(Tag)
      id={789}
      key="1991-2"
      onClose={[Function]}
      updateDisable={true}
    />
  </div>
  <div
    className="card_options --flex"
  >
    <div
      className="loading"
    >
      <Icon
        className="edit iconHover"
        name="edit"
        onClick={[Function]}
        pointer={true}
      />
    </div>
    <div
      className="loading"
    >
      <Icon
        className="close iconHover"
        name="close"
        onClick={[Function]}
        pointer={true}
      />
    </div>
  </div>
</div>
`;
